zkevm-circuits/src/witness/tx.rs

```diff
@@ -307,52 +318,448 @@ impl Transaction {
             })
             .collect()
     }
-}
 
-impl Encodable for Transaction {
-    fn rlp_append(&self, s: &mut RlpStream) {
-        s.begin_list(9);
-        s.append(&Word::from(self.nonce));
-        s.append(&self.gas_price);
-        s.append(&Word::from(self.gas));
-        if let Some(addr) = self.callee_address {
-            s.append(&addr);
+    pub(crate) fn gen_rlp_witness<F: Field>(
+        &self,
+        is_hash: bool,
+        challenges: &Challenges<Value<F>>,
+    ) -> Vec<RlpFsmWitnessRow<F>> {
+        let (rlp_bytes, format) = if is_hash {
+            (
+                self.rlp_signed.clone(),
+                match self.tx_type {
+                    TxTypes::Eip155 => TxHashEip155,
+                    TxTypes::PreEip155 => TxHashPreEip155,
+                    TxTypes::Eip1559 => TxHashEip1559,
+                    TxTypes::L1Msg => L1MsgHash,
+                    _ => unimplemented!(),
+                },
+            )
         } else {
-            s.append(&"");
+            (
+                self.rlp_unsigned.clone(),
+                match self.tx_type {
+                    TxTypes::Eip155 => TxSignEip155,
+                    TxTypes::PreEip155 => TxSignPreEip155,
+                    TxTypes::Eip1559 => TxSignEip1559,
+                    _ => unimplemented!(),
+                },
+            )
+        };
+
+        let tx_id = self.id as u64;
+        let mut witness = vec![];
+        let rom_table = format.rom_table_rows();
+        let keccak_rand = challenges.keccak_input();
+        let word_rand = challenges.evm_word();
+        let rlp_bytes_rlc = rlp_bytes
+            .iter()
+            .scan(Value::known(F::zero()), |rlc, &byte| {
+                *rlc = *rlc * keccak_rand + Value::known(F::from(byte as u64));
+
+                Some(*rlc)
+            })
+            .collect::<Vec<_>>();
+        let mut cur = SmState {
+            tag: rom_table[0].tag,
+            state: DecodeTagStart,
+            tag_idx: 0,
+            tag_length: 0,
+            tag_value_acc: Value::known(F::zero()),
+            byte_idx: 0,
+            depth: 0,
+        };
+        // TODO: what's the reason that cur_rom_row is of type `vector`.
+        // At the beginning of parsing tag we actually do not know the next tag
+        // only after we finished parsing the current tag, then we can know
```

Comment:
What do you mean by this? I didn't understand..

Reply:
refined in the https://github.com/scroll-tech/zkevm-circuits/pull/488/commits/f3f656e20de67c5028dbb1ce1ea0ced30b35c6b2. The reason that cur_rom_row is of type `vector` is when we are decoding vectors we cannot decide the next tag is `EndVector` or not. After we parsed the current tag, if the remaining bytes in this layer is zero, then we know the next tag is `EndVector`.